From 2735a973b00511e4db320eaa02d78634802c8624 Mon Sep 17 00:00:00 2001
From: Jesper Brouer <brouer@redhat.com>
Date: Thu, 14 Aug 2014 16:22:10 -0400
Subject: [net] netfilter: synproxy target: restrict to INPUT/FORWARD

Message-id: <20140814162209.11343.41843.stgit@dragon>
Patchwork-id: 87738
O-Subject: [RHEL7 net PATCH] netfilter: SYNPROXY target: restrict to INPUT/FORWARD
Bugzilla: 1081339
RH-Acked-by: Jiri Pirko <jpirko@redhat.com>
RH-Acked-by: Daniel Borkmann <dborkman@redhat.com>
RH-Acked-by: Nikolay Aleksandrov <nikolay@redhat.com>
RH-Acked-by: Florian Westphal <fwestpha@redhat.com>

Bugzilla: 1081339
Brew: http://brewweb.devel.redhat.com/brew/taskinfo?taskID=7844909

Backport of upstream commit: f01b3926ee645974f549f4a6921268142047717c

 commit f01b3926ee645974f549f4a6921268142047717c
 Author: Patrick McHardy <kaber@trash.net>
 Date:   Sun Dec 8 16:52:31 2013 +0000

    netfilter: SYNPROXY target: restrict to INPUT/FORWARD

    Fix a crash in synproxy_send_tcp() when using the SYNPROXY target in the
    PREROUTING chain caused by missing routing information.

    Reported-by: Nicki P. <xastx@gmx.de>
    Signed-off-by: Patrick McHardy <kaber@trash.net>
    Signed-off-by: Pablo Neira Ayuso <pablo@netfilter.org>

Signed-off-by: Jesper Dangaard Brouer <brouer@redhat.com>
Signed-off-by: Jiri Benc <jbenc@redhat.com>
Signed-off-by: Jarod Wilson <jarod@redhat.com>

diff --git a/net/ipv4/netfilter/ipt_SYNPROXY.c b/net/ipv4/netfilter/ipt_SYNPROXY.c
index f13bd91..a313c3f 100644
--- a/net/ipv4/netfilter/ipt_SYNPROXY.c
+++ b/net/ipv4/netfilter/ipt_SYNPROXY.c
@@ -423,6 +423,7 @@ static void synproxy_tg4_destroy(const struct xt_tgdtor_param *par)
 static struct xt_target synproxy_tg4_reg __read_mostly = {
  .name  = "SYNPROXY",
  .family  = NFPROTO_IPV4,
+ .hooks  = (1 << NF_INET_LOCAL_IN) | (1 << NF_INET_FORWARD),
  .target  = synproxy_tg4,
  .targetsize = sizeof(struct xt_synproxy_info),
  .checkentry = synproxy_tg4_check,
diff --git a/net/ipv6/netfilter/ip6t_SYNPROXY.c b/net/ipv6/netfilter/ip6t_SYNPROXY.c
index f78f41a..a0d1727 100644
--- a/net/ipv6/netfilter/ip6t_SYNPROXY.c
+++ b/net/ipv6/netfilter/ip6t_SYNPROXY.c
@@ -446,6 +446,7 @@ static void synproxy_tg6_destroy(const struct xt_tgdtor_param *par)
 static struct xt_target synproxy_tg6_reg __read_mostly = {
  .name  = "SYNPROXY",
  .family  = NFPROTO_IPV6,
+ .hooks  = (1 << NF_INET_LOCAL_IN) | (1 << NF_INET_FORWARD),
  .target  = synproxy_tg6,
  .targetsize = sizeof(struct xt_synproxy_info),
  .checkentry = synproxy_tg6_check,
-- 
1.7.1